File: packages/uns-cli/templates/default/src/examples/data-example.ts
/**
 * Change this file according to your specifications and rename it to index.ts
 */
import { UnsProxyProcess, ConfigFile, logger, type IUnsMessage, UnsAttribute } from "@uns-kit/core";
import { PhysicalMeasurements } from "@uns-kit/core/uns/uns-measurements.js";
import { UnsPacket } from "@uns-kit/core/uns/uns-packet.js";
import { UnsTopics } from "@uns-kit/core/uns/uns-topics.js";
import { ObjectTypes } from "@uns-kit/core/uns/uns-object.js";
import { EnergyResourceAttributes } from "@uns-kit/core/uns/uns-attributes.js";

/**
 * Load the configuration from a file.
 * On the server, this file is provided by the `uns-datahub-controller`.
 * In the development environment, you are responsible for creating and maintaining this file and its contents.
 */
const config = await ConfigFile.loadConfig();

/**
 * Connect to input and output brokers
 */
const unsProxyProcess = new UnsProxyProcess(config.infra.host!, {processName: config.uns.processName!});
const mqttInput = await unsProxyProcess.createUnsMqttProxy((config.input?.host)!, "templateUnsRttInput", config.uns.instanceMode!, config.uns.handover!, {
  mqttSubToTopics: ["raw/#"],
});
const mqttOutput = await unsProxyProcess.createUnsMqttProxy((config.output?.host)!, "templateUnsRttOutput", config.uns.instanceMode!, config.uns.handover!, { publishThrottlingDelay: 1000});


/**
 * Event listener for input events.
 * Transform an input message and publish it with publishMqttMessage function.
 */
mqttInput.event.on("input", async (event) => {
  try {
    if (event.topic === "raw/data") {
      const values = event.message.split(",");
      const [countRaw, timestampRaw, sensorRaw] = values;
      if (!countRaw || !timestampRaw || !sensorRaw) {
        logger.warn(`Skipping malformed raw/data payload: ${event.message}`);
        return;
      }
      const numberValue = Number.parseFloat(countRaw);
      const eventDate = new Date(Number.parseInt(timestampRaw, 10));
      const sensorValue = Number.parseFloat(sensorRaw);
      const time = UnsPacket.formatToISO8601(eventDate);

      const dataGroup = "sensor";

      const message: IUnsMessage = {
        data: { dataGroup, time, value: numberValue, uom: PhysicalMeasurements.None },
      };
      const topic: UnsTopics = "enterprise/site/area/line/";
      const currentPacket = await UnsPacket.unsPacketFromUnsMessage(message);
      mqttOutput.publishMqttMessage({
        topic,
        asset:"asset",
        objectType: ObjectTypes.EnergyResource,
        objectId: "main",
        attribute: EnergyResourceAttributes.Current,
        description: "Simulated current sensor value",
        tags: [],
        packet: currentPacket
      });

      const sensorMessage: IUnsMessage = {
        data: { dataGroup, time, value: sensorValue, uom: PhysicalMeasurements.Celsius },
      };
      const sensorPacket = await UnsPacket.unsPacketFromUnsMessage(sensorMessage);
      mqttOutput.publishMqttMessage({
        topic,
        asset:"asset",
        objectType: ObjectTypes.EnergyResource,
        objectId: "main",
        attribute: EnergyResourceAttributes.Voltage,
        description: "Simulated voltage sensor value",
        tags: [],
        packet: sensorPacket
      });
    }
  } catch (error) {
    const reason = error instanceof Error ? error : new Error(String(error));
    logger.error(`Error publishing message to MQTT: ${reason.message}`);
    throw reason;
  }
});

File: packages/uns-py/examples/data_example.py
import asyncio
from datetime import datetime, timedelta, timezone
from pathlib import Path

from uns_kit import ConfigFile, TopicBuilder, UnsMqttClient, UnsProcessParameters, UnsProxyProcess
from uns_kit.packet import isoformat


async def main() -> None:
    cfg_path = Path("config.json")
    cfg = ConfigFile.load_config(cfg_path) if cfg_path.exists() else {"infra": {"host": "localhost"}, "uns": {"processName": "uns-process"}}
    infra = cfg.get("infra") or {}
    uns = cfg.get("uns") or {}
    host = infra.get("host") or "localhost"
    port = infra.get("port")
    process_name = uns.get("processName") or "uns-process"
    process = UnsProxyProcess(host, UnsProcessParameters(process_name=process_name))
    await process.start()
    print("[data-example] Process client connected.")

    mqtt_output = await process.create_mqtt_proxy("py-output")
    print("[data-example] Output proxy connected.")
    mqtt_input = UnsMqttClient(
        host,
        port=port,
        username=infra.get("username") or None,
        password=infra.get("password") or None,
        tls=bool(infra.get("tls")),
        client_id=None,
        topic_builder=TopicBuilder(process_name=process_name),
        instance_name="py-input",
        subscriber_active=True,
    )
    await mqtt_input.connect()
    print("[data-example] Input client connected, subscribing to raw/#")

    topic = "enterprise/site/area/line/"
    asset = "asset"
    asset_description = "Sample asset"
    object_type = "energy-resource"
    object_id = "main"

    try:
        async for msg in mqtt_input.resilient_messages("raw/#"):
            if msg.topic.value != "raw/data":
                continue
            payload = msg.payload.decode()
            values = payload.split(",")
            if len(values) < 3:
                print(f"Skipping malformed raw/data payload: {payload}")
                continue

            count_raw, timestamp_raw, sensor_raw = values[:3]
            try:
                number_value = float(count_raw)
                event_time = datetime.fromtimestamp(int(timestamp_raw) / 1000, tz=timezone.utc)
                sensor_value = float(sensor_raw)
            except ValueError:
                print(f"Skipping malformed raw/data payload: {payload}")
                continue

            interval_start = isoformat(event_time - timedelta(seconds=1))
            interval_end = isoformat(event_time)
            time = isoformat(event_time)

            await mqtt_output.publish_mqtt_message(
                {
                    "topic": topic,
                    "asset": asset,
                    "assetDescription": asset_description,
                    "objectType": object_type,
                    "objectId": object_id,
                    "attributes": [
                        {
                            "attribute": "current",
                            "description": "Simulated current sensor value",
                            "data": {
                                "dataGroup": "sensor",
                                "time": time,
                                "intervalStart": interval_start,
                                "intervalEnd": interval_end,
                                "value": number_value,
                                "uom": "A",
                            },
                        },
                        {
                            "attribute": "voltage",
                            "description": "Simulated voltage sensor value",
                            "data": {
                                "dataGroup": "sensor",
                                "time": time,
                                "intervalStart": interval_start,
                                "intervalEnd": interval_end,
                                "value": sensor_value,
                                "uom": "V",
                            },
                        },
                    ],
                }
            )
            print(f"[data-example] Published transformed attributes for ts={time}")
    except KeyboardInterrupt:
        pass
    finally:
        await mqtt_input.close()
        await mqtt_output.close()
        await process.stop()


if __name__ == "__main__":
    asyncio.run(main())
